tests/test_storage.py

```diff
@@ -0,0 +1,235 @@
+# -*- coding: utf-8 -*-
+"""
+This is a unit test for representing storage in the MESSAGEix model, and
+testing the functionality of storage equations. The workflow is as follows:
+    - building a stylized MESSAGEix model
+    - adding seasonality and modifying parameters accordingly
+    - adding storage implementation
+    - testing storage functionality and equations
+
+"""
+
+from message_ix import Scenario
+
+# A flag to show if storage variables/parameters are added to ixmp java backend
+storage_in_java = False
+
+
+# A function for generating a simple MESSAGEix model
+def model_setup(scen, years):
+    scen.add_set('node', 'node')
+    scen.add_set('commodity', 'electr')
+    scen.add_set('level', 'level')
+    scen.add_set('year', years)
+    scen.add_set('type_year', years)
+    scen.add_set('technology', ['tec1', 'tec2'])
+    scen.add_set('mode', 'mode')
+    output_specs = ['node', 'electr', 'level', 'year', 'year']
+    dict_var_cost = {'tec1': 1, 'tec2': 2}
+
+    for yr in years:
+        scen.add_par('demand', ['node', 'electr', 'level', yr, 'year'],
+                     1, 'GWa')
+        for t in dict_var_cost.keys():
+            tec_specs = ['node', t, yr, yr, 'mode']
+            scen.add_par('output', tec_specs + output_specs, 1, 'GWa')
+            scen.add_par('var_cost',
+                         tec_specs + ['year'], dict_var_cost[t], 'USD/GWa')
+
+
+# A function for adding sub-annual time steps to a MESSAGEix model
+def add_seasonality(scen, seasons_dict):
+    scen.add_set('time', list(set(seasons_dict.keys())))
+    scen.add_set('lvl_temporal', 'season')
+    for h in seasons_dict.keys():
+        scen.add_set('map_temporal_hierarchy', ['season', h, 'year'])
+        scen.add_par('duration_time', [h], seasons_dict[h], '-')
+
+
+# A function for modifying model parameters after adding sub-annual time steps
+def year_to_time(scen, parname, dict_par, filters=None):
+    if filters:
+        old = scen.par(parname, filters)
+    else:
+        old = scen.par(parname)
+    scen.remove_par(parname, old)
+    time_idx = [x for x in scen.idx_names(parname) if 'time' in x]
+    for h in dict_par.keys():
+        new = old.copy()
+        for index in time_idx:
+            new[index] = h
+        new['value'] = dict_par[h] * old['value']
+        scen.add_par(parname, new)
+
+
+# A function for initiating sets/parameters required for storage equations
+# NOTICE: this function can be deleted when storage parameters/sets are added
+# to ixmp java backend
+def init_storage(scen):
```

Comment:
So we actually have support for this already in the code base~! Check out [this line](https://github.com/iiasa/message_ix/blob/master/message_ix/core.py#L26). I would suggest putting this there (feel free to stop by and I can show you how). Then when it gets into `ixmp`, we can take it out again. 

On the other hand, I would also be happy if we wanted to move this work all into python, but I'm not sure that's in the scope of this PR =)

Reply:
Thanks @Matt for the explanation!